cajita/unit_test/tstSparseIndexSpace.hpp

```diff
@@ -0,0 +1,636 @@
+    int capacity = size_per_dim * size_per_dim;
+    float rehash_factor = 1.5;
+    SparseIndexSpace<TEST_EXECSPACE> sis( size, capacity, rehash_factor );
+
+    auto cbd = sis.cell_bits_per_tile_dim;
+    EXPECT_EQ( cbd, 2 );
+
+    auto cnd = sis.cell_num_per_tile_dim;
+    EXPECT_EQ( cnd, 4 );
+
+    auto cmd = sis.cell_mask_per_tile_dim;
+    EXPECT_EQ( cmd, 3 );
+
+    auto cbt = sis.cell_bits_per_tile;
+    EXPECT_EQ( cbt, 6 );
+
+    auto cnt = sis.cell_num_per_tile;
+    EXPECT_EQ( cnt, 64 );
+
+    Kokkos::View<int***, TEST_DEVICE> iid_res( "insert_id", size[0], size[1],
+                                               size[2] );
+    Kokkos::View<int***, TEST_DEVICE> qid_res( "query_id", size[0], size[1],
+                                               size[2] );
+
+    Kokkos::parallel_for(
+        Kokkos::RangePolicy<TEST_EXECSPACE>( 0, size_per_dim ),
+        KOKKOS_LAMBDA( int i ) {
+            for ( int j = 0; j < size_per_dim; j++ )
+                for ( int k = 0; k < size_per_dim; k++ )
+                {
+                    iid_res( i, j, k ) = sis.insertCell( i, j, k );
+                }
+        } );
+
+    Kokkos::parallel_for(
+        Kokkos::RangePolicy<TEST_EXECSPACE>( 0, size_per_dim ),
+        KOKKOS_LAMBDA( int i ) {
+            for ( int j = 0; j < size_per_dim; j++ )
+                for ( int k = 0; k < size_per_dim; k++ )
+                {
+                    qid_res( i, j, k ) = sis.queryCell( i, j, k );
+                }
+        } );
+
+    bool test[size_per_dim * size_per_dim * size_per_dim];
+    auto iid_mirror =
+        Kokkos::create_mirror_view_and_copy( Kokkos::HostSpace(), iid_res );
+    auto qid_mirror =
+        Kokkos::create_mirror_view_and_copy( Kokkos::HostSpace(), qid_res );
+
+    for ( int i = 0; i < size[0]; i++ )
+        for ( int j = 0; j < size[1]; j++ )
+            for ( int k = 0; k < size[2]; k++ )
+                test[i * size[1] * size[2] + j * size[2] + k] = false;
+
+    for ( int i = 0; i < size[0]; i++ )
+        for ( int j = 0; j < size[1]; j++ )
+            for ( int k = 0; k < size[2]; k++ )
+                test[qid_mirror( i, j, k )] = true;
+
+    for ( int i = 0; i < size[0]; i++ )
+        for ( int j = 0; j < size[1]; j++ )
+            for ( int k = 0; k < size[2]; k++ )
+            {
+                EXPECT_EQ( qid_mirror( i, j, k ), iid_mirror( i, j, k ) );
+                EXPECT_EQ( test[i * size[2] * size[1] + j * size[2] + k],
+                           true );
+            }
+
+    constexpr int total_tile_num =
+        size_tile_per_dim * size_tile_per_dim * size_tile_per_dim;
+
+    uint32_t cap = sis.capacity();
+    EXPECT_EQ( cap >= total_tile_num, true );
+
+    auto s = sis.size();
+    EXPECT_EQ( s, total_tile_num );
+
+    uint32_t new_set_cap = total_tile_num * 10;
+    sis.reserve( new_set_cap );
+    uint32_t new_cap = sis.capacity();
+    EXPECT_EQ( new_cap >= new_set_cap, true );
```

Comment:
The capacity is handled by the `capacity` interface from the Kokkos Unordered map. It can only ensure the lower bound but not the exact capacity value.